Enhancing The Update Framework (Python - TUF) README

shubhusion opened the pull request:
<!--
Before submitting a pull request:
  * Run linter and tests locally: Use "tox"
  * Ensure your commits are signed-off-by: Use "commit --signoff"
  * Make sure new code has tests and is documented
  * For more info, see docs/CONTRIBUTING.rst

Once commits are signed off and tested, describe the purpose and contents
of the pull request below.
-->

**Description of the changes being introduced by the pull request**:

1. **Table of Contents**: Added a Table of Contents to improve navigation and accessibility within the README.
   
2. **Logo Color**: Changed the logo color from black to white to enhance visibility and improve the overall aesthetic of the document.
   
3. **Additional Badges**: Included additional badges to highlight key aspects such as version, license, tech-stack, and community support, providing users with quick access to essential information.
   
4. **Back to Top Button**: Added a "Back to Top" button at the bottom of the README to facilitate easy navigation within the document.
   
5. **Installation Section**: Added an Installation section to provide users with clear guidance on how to install the software.

6. **Contact Section Enhancement**: Separated the Mailing List and CNCF Slack links for better readability and organization within the Contact section.

7. **CNCF Logo Addition**: Added the CNCF logo at the bottom of the README to acknowledge the project's affiliation with the Cloud Native Computing Foundation.

**Before:**
![image](https://github.com/theupdateframework/python-tuf/assets/96301987/cebd9a21-26b3-4729-adeb-a5746dc916c9)


**After:**
![image](https://github.com/theupdateframework/python-tuf/assets/96301987/3c9e83ac-67a1-461b-988f-bd79d9c7de35)


Fixes #2654

Reviewers: 

@trishankatdatadog @JustinCappos

dependabot[bot] committed:
build(deps): bump the test-and-lint-dependencies group with 2 updates

Bumps the test-and-lint-dependencies group with 2 updates: [coverage](https://github.com/nedbat/coveragepy) and [ruff](https://github.com/astral-sh/ruff).

Updates `coverage` from 7.5.2 to 7.5.3
- [Release notes](https://github.com/nedbat/coveragepy/releases)
- [Changelog](https://github.com/nedbat/coveragepy/blob/master/CHANGES.rst)
- [Commits](https://github.com/nedbat/coveragepy/compare/7.5.2...7.5.3)

Updates `ruff` from 0.4.5 to 0.4.7
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/v0.4.5...v0.4.7)

---
updated-dependencies:
- dependency-name: coverage
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

dependabot[bot] committed:
build(deps): bump the test-and-lint-dependencies group with 2 updates

Bumps the test-and-lint-dependencies group with 2 updates: [coverage](https://github.com/nedbat/coveragepy) and [ruff](https://github.com/astral-sh/ruff).

Updates `coverage` from 7.5.2 to 7.5.3
- [Release notes](https://github.com/nedbat/coveragepy/releases)
- [Changelog](https://github.com/nedbat/coveragepy/blob/master/CHANGES.rst)
- [Commits](https://github.com/nedbat/coveragepy/compare/7.5.2...7.5.3)

Updates `ruff` from 0.4.5 to 0.4.7
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/v0.4.5...v0.4.7)

---
updated-dependencies:
- dependency-name: coverage
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>

h4l0gen committed:
Updating Contributing guidelines and copy-pastable code (#2642)

* Make commands easier to copy

Signed-off-by: h4l0gen <ks3913688@gmail.com>
Signed-off-by: Kapil Sharma <ks3913688@gmail.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

h4l0gen committed:
repository:  handle online key changes (#2650)

* repository: Handle online key change situations in do_snapshot() and do_timestamp():
  always create a new version if current version is not correctly signed
* remove expectedFailure marks from the related tests

Signed-off-by: h4l0gen <ks3913688@gmail.com>
Signed-off-by: Kapil Sharma <ks3913688@gmail.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion commented:
@trishankatdatadog @JustinCappos please review this PR for the issue #2654

dependabot[bot] committed:
build(deps): bump ruff in the test-and-lint-dependencies group

Bumps the test-and-lint-dependencies group with 1 update: [ruff](https://github.com/astral-sh/ruff).

Updates `ruff` from 0.4.7 to 0.4.8
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/v0.4.7...v0.4.8)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

dependabot[bot] committed:
build(deps): bump ruff in the test-and-lint-dependencies group

Bumps the test-and-lint-dependencies group with 1 update: [ruff](https://github.com/astral-sh/ruff).

Updates `ruff` from 0.4.7 to 0.4.8
- [Release notes](https://github.com/astral-sh/ruff/releases)
- [Changelog](https://github.com/astral-sh/ruff/blob/main/CHANGELOG.md)
- [Commits](https://github.com/astral-sh/ruff/compare/v0.4.7...v0.4.8)

---
updated-dependencies:
- dependency-name: ruff
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: test-and-lint-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

jku commented:
> 2\. **Logo Color**: Changed the logo color from black to white to enhance visibility and improve the overall aesthetic of the document.

This change makes the logo invisible with default GitHub settings (You are using dark theme on GitHub so see the opposite effect).

shubhusion commented:
> > 2. **Logo Color**: Changed the logo color from black to white to enhance visibility and improve the overall aesthetic of the document.
> 
> This change makes the logo invisible with default GitHub settings (You are using dark theme on GitHub so see the opposite effect).

To ensure the logo is visible in both light and dark themes, a solution that automatically switches the logo based on the user's selected theme preference can be implemented. This strategy involves utilizing two different logo versions (one for light theme and one for a dark theme) and utilizing HTML and CSS to manage the theme-based switching.

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
added assets

Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

dependabot[bot] committed:
build(deps): bump the dependencies group with 2 updates (#2657)

Bumps the dependencies group with 2 updates: [cryptography](https://github.com/pyca/cryptography) and [securesystemslib[crypto,pynacl]](https://github.com/secure-systems-lab/securesystemslib).

Updates `cryptography` from 42.0.7 to 42.0.8
- [Changelog](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/42.0.7...42.0.8)

Updates `securesystemslib[crypto,pynacl]` from 1.0.0 to 1.1.0
- [Release notes](https://github.com/secure-systems-lab/securesystemslib/releases)
- [Changelog](https://github.com/secure-systems-lab/securesystemslib/blob/main/CHANGELOG.md)
- [Commits](https://github.com/secure-systems-lab/securesystemslib/compare/v1.0.0...v1.1.0)

---
updated-dependencies:
- dependency-name: cryptography
  dependency-type: direct:production
  update-type: version-update:semver-patch
  dependency-group: dependencies
- dependency-name: securesystemslib[crypto,pynacl]
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
added assets

Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Update README.md

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>
Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion commented:
@jku it works for both light and dark theme now. please have a look

jku commented on README.md:
```diff
@@ -1,113 +1,143 @@
-# <img src="https://cdn.rawgit.com/theupdateframework/artwork/3a649fa6/tuf-logo.svg" height="100" valign="middle" alt="TUF"/> A Framework for Securing Software Update Systems
+<picture>
+  <source media="(prefers-color-scheme: dark)" srcset="https://github.com/shubhusion/python-tuf/blob/develop/tuf-horizontal-white.png">
+  <source media="(prefers-color-scheme: light)" srcset="https://github.com/shubhusion/python-tuf/blob/develop/tuf-horizontal-black.png">
+  <img alt="Shows an illustrated sun in light mode and a moon with stars in dark mode." src="https://github.com/shubhusion/python-tuf/blob/develop/tuf-horizontal-black.png">
+</picture>
```
* alt text is incorrect
* I don't think you need to specify `source` for light theme: the default `img` is enough
* I think linking to pngs in cncf/artwork project would be totally fine, even preferable: if you do want to upload the pngs to this project then
   * please don't put them directly in the root directory
   * `https://github.com/shubhusion/python-tuf/` is not ok as url

jku commented on README.md:
```diff
@@ -1,113 +1,143 @@
 [![CII](https://bestpractices.coreinfrastructure.org/projects/1351/badge)](https://bestpractices.coreinfrastructure.org/projects/1351)
 [![PyPI](https://img.shields.io/pypi/v/tuf)](https://pypi.org/project/tuf/)
 [![OpenSSF Scorecard](https://api.securityscorecards.dev/projects/github.com/theupdateframework/python-tuf/badge)](https://api.securityscorecards.dev/projects/github.com/theupdateframework/python-tuf)
+[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
+[![License: Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
```
These links are already provided by github as clickable tabs just a few cm above this.

jku commented on README.md:
```diff
@@ -1,113 +1,143 @@
+
+---
+
+## Table of Contents
+
+- [Introduction](#introduction)
+- [About The Update Framework](#about-the-update-framework)
+- [Key Features](#key-features)
+- [Documentation](#documentation)
+- [Installation](#installation)
+- [Contact](#contact)
+- [Security Issues and Bugs](#security-issues-and-bugs)
+- [License](#license)
+- [Acknowledgements](#acknowledgements)
```
GitHub already provides an automatic  TOC in the tab bar now that you've added headers. I have to wonder if that would be enough (considering that one requires 0 maintenance)

jku commented on README.md:
```diff
@@ -1,113 +1,143 @@
-**reference implementation** written in Python. It is intended to conform to
-version 1.0 of the [TUF
-specification](https://theupdateframework.github.io/specification/latest/).
+---
```
Now that you are using headers (good idea), I think these horizontal lines are just clutter: I think most of them can just be removed

jku commented on README.md:
```diff
@@ -1,113 +1,143 @@
+</div>
+
+
+---
+
+[![](https://img.shields.io/badge/Move%20to%20Top-%E2%86%91%20Back%20to%20Top-blue)](#a-framework-for-securing-software-update-systems)
+
```
This duplicates functionality people already have in their browsers, except now we have to maintain that functionality. Let's not add things like this.

jku reviewed (changes requested):
I've not looked at the text content changes(would appreciate someone else having a look). 

Structural changes look nice (like use of actual headers) but I left comments on the urls etc

trishankatdatadog commented:
> @trishankatdatadog @JustinCappos please review this PR for the issue #2654

Sorry, I'm running out of B/W... please try someone else? @jku has been massively helpful here, thanks!

lukpueh commented:
@jku already left a thorough review here. From what I see, the key improvement of this PR is to use the correct markup for the section headers. @shubhusion, do you plan to address the review comments?

shubhusion commented:
> @jku already left a thorough review here. From what I see, the key improvement of this PR is to use the correct markup for the section headers. @shubhusion, do you plan to address the review comments?

@lukpueh yeah I will make changes as specified

shubhusion committed:
Fixed Readme

Signed-off-by: shubhusion <shubham27.sharma03@gmail.com>

shubhusion committed:
Merge branch 'develop' into develop

Signed-off-by: SHUBHAM SHARMA  <96301987+shubhusion@users.noreply.github.com>

jku commented on README.md:
```diff
@@ -7,107 +12,91 @@
+
+---
+
+## <div align="center"> The Update Framework (TUF) is a CNCF graduated project </div>
+
+<div align="center">
+  <a href="https://www.cncf.io/">
+    <picture>
+      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/shubhusion/python-tuf/blob/develop/cncf-white.png">
+      <source media="(prefers-color-scheme: light)" srcset="https://github.com/shubhusion/python-tuf/blob/develop/cncf-black.png">
+      <img src="https://github.com/shubhusion/python-tuf/blob/develop/cncf-black.png" height="150" alt="TUF" style="background-color:white"/>
+    </picture>
+  </a>
+</div>
```
* links to "https://github.com/shubhusion/ are still not ok
* I also wonder if CNCF really needs a second level title and a huge logo in addition to the acknowledgement it already had. We could just tweak the previous CNCF reference in "Acknowledgements" to include the mention of being a graduated project?

jku reviewed (changes requested):
* There are now unused pngs included in the PR: please remove those
* There are some issues with the CNCF section at the bottom, left a comment there

**Note to PR merger**: If the pngs are in any commit in the PR, this should be squash merged to prevent them from being part of git history

jku commented:
I'm closing this since no action in 2 months. 

Please reopen if you get to it at a later date
